langchain4j-watsonx/src/main/java/dev/langchain4j/model/watsonx/WatsonxChatModel.java

```
@@ -0,0 +1,168 @@
+import dev.langchain4j.exception.ContentFilteredException;
+import dev.langchain4j.model.ModelProvider;
+import dev.langchain4j.model.chat.Capability;
+import dev.langchain4j.model.chat.ChatModel;
+import dev.langchain4j.model.chat.listener.ChatModelListener;
+import dev.langchain4j.model.chat.request.ChatRequest;
+import dev.langchain4j.model.chat.request.ChatRequestParameters;
+import dev.langchain4j.model.chat.response.ChatResponse;
+import dev.langchain4j.model.output.FinishReason;
+import dev.langchain4j.model.output.TokenUsage;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A {@link ChatModel} implementation that integrates IBM watsonx.ai with LangChain4j.
+ * <p>
+ * <b>Example usage:</b>
+ *
+ * <pre>{@code
+ * ChatModel chatModel = WatsonxChatModel.builder()
+ *     .url("https://...") // or use CloudRegion
+ *     .apiKey("...")
+ *     .projectId("...")
+ *     .modelName("ibm/granite-3-8b-instruct")
+ *     .maxOutputTokens(0)
+ *     .temperature(0.7)
+ *     .build();
+ * }</pre>
+ *
+ */
+public class WatsonxChatModel extends WatsonxChat implements ChatModel {
+
+    private WatsonxChatModel(Builder builder) {
+        super(builder);
+    }
+
+    @Override
+    public ChatResponse doChat(ChatRequest chatRequest) {
+
+        validate(chatRequest.parameters());
+
+        List<ToolSpecification> toolSpecifications = getOrDefault(
+                chatRequest.parameters().toolSpecifications(), defaultRequestParameters.toolSpecifications());
+
+        List<ChatMessage> messages =
+                chatRequest.messages().stream().map(Converter::toChatMessage).collect(toCollection(ArrayList::new));
+
+        List<Tool> tools = nonNull(toolSpecifications) && toolSpecifications.size() > 0
+                ? toolSpecifications.stream().map(Converter::toTool).toList()
+                : null;
+
+        if (isThinkingActivable(chatRequest.messages(), toolSpecifications)) messages.add(THINKING);
```

Shouldn't thinking be activated by the user explicitly enabling it?

---

The problem here (as you can see from the code) is that today, watsonx.ai has some limitations when the thinking part is enabled. First of all, today only Granite is enabled for thinking, and second, when it is enabled, no SystemMessage or Tools can be used. What happens today is that if one of these is sent together with the `TextChatMessageControl` message, the thinking part will be ignored. I have added warning logs to notify the developer. Looking at the code, maybe an exception would be better.

---

AFAIU current implementation always enables thinking when there is no system message and tools. But what if user does not want thinking at all? Is there a way to disable it? As a user, I would expect to have some switch to enable/disable thinking that I can set when creating `WatsonxChatModel` or when creating `ChatReqeust`

---

To enable thinking, the developer must pass the `ExtractionTags` to the builder.

```java
boolean isThinkingActivable(List<ChatMessage> messages, List<ToolSpecification> tools) {
        if (isNull(tags)) return false;
```

If no tags are present, thinking will always be disabled.

---

To the switch is
```java
WatsonxChatModel.builder()
            ...
            .thinking(ExtractionTags.of("think", "response"))
            .build();
```

---

Another point to consider is whether logging a warning is sufficient or whether it would be better to send an exception if the developer enables the thinking part and sends one or more tools or a SystemMessage.

---

@andreadimaio got it, thanks for the explanation!

Yes, I would rather throw an exception, otherwise it can become confusing (logs not enabled, etc)

---

@dliubarskyi, please take a look at the latest commit.

---

Looks good, although I would use more specific `UnsupportedFeatureException` instead of `LangChain4jException`

---

I will incorporate this change into the new PR.